### PR TITLE
Inter-Foil Distance Spatial Bias: Aerodynamic Coupling Routing

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The spatial_bias MLP currently routes tokens using `[x, y, curvature_proxy, dist_to_nearest_foil, gap, stagger]`. The **distance-to-nearest-foil** feature treats proximity to foil-1 and foil-2 identically. For tandem configurations, what matters architecturally is how close each node is to the **opposite** foil — nodes in the narrow inter-foil gap experience strong aerodynamic coupling (pressure interference, wake effects), while nodes on the outer surfaces experience weaker interaction.

By adding `foil2_dist_feat` (log-scaled min distance to foil-2, computed from raw DSDF channels 4:8) as a **7th input** to the spatial_bias MLP, we give the routing mechanism explicit awareness of aerodynamic coupling strength. High-coupling nodes (in the gap) and low-coupling nodes (outer surfaces) will be routed to different slices, enabling the model to develop specialized representations for each.

For single-foil samples, foil-2 DSDF channels are all zero → `foil2_dist_feat = 0.0` for all nodes → the new input cleanly degrades to a no-op. For tandem samples, gap nodes get a distinct non-zero signal, giving the model a geometrically-grounded routing criterion.

**This extends the largest single improvement in the programme** — `gap_stagger_spatial_bias` (PR #2130, p_tan -3.0%) — by adding a per-node inter-foil coupling signal to complement the existing per-sample gap/stagger scalars. Those capture global tandem geometry; this captures local coupling at each mesh node.

**Reference:** `foil2_dist_feat` already exists in the codebase (computed when `--foil2_dist` is used, train.py ~line 1679), but is not currently wired into the spatial_bias MLP.

### Step 1: Add CLI flag

```python
# In TrainConfig dataclass:
interfoil_spatial_bias: bool = False   # add foil2_dist to spatial_bias inputs (requires gap_stagger_spatial_bias)
```

### Step 2: Extend spatial_bias input dim

In `Transolver.__init__`, where `spatial_bias_input_dim` is determined (around line 774):

```python
if gap_stagger_spatial_bias:
    if interfoil_spatial_bias:
        spatial_bias_input_dim = 7   # [x, y, curv, dist, gap, stagger, foil2_dist]
    else:
        spatial_bias_input_dim = 6   # [x, y, curv, dist, gap, stagger]
else:
    spatial_bias_input_dim = 4   # [x, y, curv, dist]
```

Also store `self.interfoil_spatial_bias = interfoil_spatial_bias` on the model.

### Step 3: Zero-init the new column

After the existing zero-init for gap/stagger columns (lines 779–783):

```python
if gap_stagger_spatial_bias and interfoil_spatial_bias:
    with torch.no_grad():
        for block in self.blocks:
            block.spatial_bias[0].weight[:, 6:].zero_()
```

### Step 4: Extend x and wire into raw_xy

In the training loop (~line 1679), after computing `dist_feat`, always compute `foil2_dist_feat_sb`:

```python
raw_dsdf_f2 = x_raw[:, :, 6:10]  # foil-2 DSDF channels (0.0 for single-foil)
foil2_dist_feat_sb = torch.log1p(raw_dsdf_f2.abs().min(dim=-1, keepdim=True).values * 10.0)
# shape [B, N, 1]; 0.0 for single-foil samples
```

Append it to x BEFORE Fourier PE, as the new 3rd extra scalar (after curv and dist_feat):
```python
# After existing: x = torch.cat([x, curv, dist_feat], dim=-1)
if cfg.interfoil_spatial_bias:
    x = torch.cat([x, curv, dist_feat, foil2_dist_feat_sb], dim=-1)
    # Now x has 27 channels before Fourier PE
else:
    x = torch.cat([x, curv, dist_feat], dim=-1)
    # Standard 26 channels before Fourier PE
```

Update `fun_dim` in model init:
```python
fun_dim = X_DIM - 2 + 2 + (1 if cfg.foil2_dist else 0) + (1 if cfg.interfoil_spatial_bias else 0) + 32
```

Do the same in the validation loop (same pattern as training loop).

In `Transolver.forward`, update raw_xy construction:

```python
if self.gap_stagger_spatial_bias:
    gap_stagger = x[:, 0:1, 22:24].expand(-1, x.shape[1], -1)   # [B, N, 2]
    if self.interfoil_spatial_bias:
        # foil2_dist is at index 26 (after curv=24, dist=25, foil2_dist=26)
        foil2_dist_sb = x[:, :, 26:27]  # [B, N, 1]
        raw_xy = torch.cat([x[:, :, :2], x[:, :, 24:26], gap_stagger, foil2_dist_sb], dim=-1)
        # [B, N, 7]: x, y, curv, dist, gap, stagger, foil2_dist
    else:
        raw_xy = torch.cat([x[:, :, :2], x[:, :, 24:26], gap_stagger], dim=-1)
else:
    raw_xy = torch.cat([x[:, :, :2], x[:, :, 24:26]], dim=-1)
```

⚠️ **Index adjustment:** The index `26` assumes `foil2_dist_feat_sb` is appended as the 3rd extra scalar (after curv=24, dist=25). If `--foil2_dist` is also enabled, adjust index accordingly (foil2_dist_sb would be at 27).

### Step 5: Run 2 seeds

```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent tanjiro --seed 42 \
  --wandb_name "tanjiro/interfoil-sb-s42" \
  --wandb_group "tanjiro/interfoil-sb" \
  --interfoil_spatial_bias \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5

# Seed 73: same command, --seed 73, --wandb_name "tanjiro/interfoil-sb-s73"
```

### Key notes

1. **foil2_dist is zero for single-foil** — their foil-2 DSDF channels (raw x[:,:,6:10]) are zero, so `log1p(0*10) = 0.0`. Clean no-op for single-foil routing.
2. **Requires `--gap_stagger_spatial_bias`** — add an assert if interfoil_spatial_bias=True but gap_stagger_spatial_bias=False.
3. **Sanity check:** Print foil2_dist_feat_sb min/max for one tandem batch and one single-foil batch. Tandem should have range ~[0, 2.5]; single-foil should be all zeros.
4. **fun_dim update is critical** — verify model init gets the correct fun_dim to avoid shape mismatches.
5. **VRAM:** Negligible overhead (~43 GB).

## Baseline

| Metric | 2-seed avg | Target to beat |
|--------|-----------|----------------|
| p_in | **13.21** | < 13.21 |
| p_oodc | **7.82** | < 7.82 |
| **p_tan** | **28.50** | **< 28.50** |
| p_re | **6.45** | < 6.45 |

**Baseline W&B runs:** 6yfv5lio (s42, p_tan=28.432), etepxvjc (s73, p_tan=28.572) — PR #2184 (DCT Freq Loss w=0.05)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent tanjiro --wandb_name "tanjiro/baseline-dct-freq" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5
```

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2194 because Closing as duplicate — tanjiro is already running the curvature loss weighting experiment in PR #2197. This inter-foil distance experiment is assigned to askeladd in PR #2195, which has the final implementation. Avoiding redundant GPU usage..

**Comment by tcapelle** [comment]
## Experiment Results — Inter-Foil Distance Spatial Bias

### Implementation
- Added `--interfoil_spatial_bias` flag
- Computes `foil2_dist_feat_sb = log1p(min_foil2_dsdf * 10)` per node, appended to x as extra scalar
- Extends spatial_bias MLP input from 6→7: [x, y, curv, dist, gap, stagger, foil2_dist]
- Zero-inits new column so initial routing is unchanged
- Cleanly degrades to 0 for single-foil samples (foil-2 DSDF channels are all zero)

### Results (2 seeds)

| Metric | Baseline (2-seed avg) | Interfoil s42 (srn2nh08) | Interfoil s73 (17l5t3we) | Avg | Delta |
|--------|----------------------|--------------------------|--------------------------|-----|-------|
| p_in | **13.21** | 13.6 | 13.1 | 13.35 | +1.1% |
| p_oodc | **7.82** | 8.0 | 7.6 | 7.80 | -0.3% |
| **p_tan** | **28.50** | 29.2 | 29.1 | **29.15** | **+2.3%** |
| p_re | **6.45** | 6.4 | 6.4 | 6.40 | -0.8% |

Both ran to epoch 147 (180 min timeout).

### Analysis

**The interfoil distance spatial bias does not improve results.** p_tan increased from 28.50 → 29.15 (+2.3%), and p_in from 13.21 → 13.35 (+1.1%). p_oodc and p_re showed marginal improvements (-0.3% and -0.8% respectively) that don't compensate.

**Why it may not have helped:** The spatial_bias MLP already receives the `dist_feat` (minimum distance to nearest foil) and gap/stagger scalars. Adding the foil-2 distance specifically may be redundant — for nodes near the gap, `dist_feat` already captures proximity to the nearest foil surface, which IS the opposite foil for gap nodes. The additional signal may not provide geometrically distinct information that the model can't already infer from the existing 6 inputs.

The zero-init ensures the new feature starts as a no-op, but 147 epochs may not be enough for the model to learn useful routing patterns from this additional channel, especially since the existing 6-feature routing is already well-optimized.

### Suggested follow-ups
1. **Signed distance rather than absolute:** Use signed foil-2 distance (positive=outside, negative=inside) to give directional coupling information
2. **Focus on loss-side improvements:** The routing is already strong; loss reformulation (curvature weighting, frequency loss) may have more headroom
